src/services/authorization.rs

```diff
@@ -1,199 +1,142 @@
 //! Authorization service.
 use std::sync::Arc;
 
+use casbin::prelude::*;
+use serde::{Deserialize, Serialize};
+use tokio::sync::RwLock;
+
 use super::user::Repository;
 use crate::errors::ServiceError;
 use crate::models::user::{UserCompact, UserId};
 
+#[derive(Debug, Clone, Serialize, Deserialize, Hash)]
 pub enum ACTION {
     AddCategory,
     DeleteCategory,
     GetSettings,
     GetSettingsSecret,
     AddTag,
     DeleteTag,
+    DeleteTorrent,
+    BanUser,
 }
 
 pub struct Service {
     user_repository: Arc<Box<dyn Repository>>,
+    casbin_enforcer: Arc<CasbinEnforcer>,
 }
 
 impl Service {
     #[must_use]
-    pub fn new(user_repository: Arc<Box<dyn Repository>>) -> Self {
-        Self { user_repository }
+    pub fn new(user_repository: Arc<Box<dyn Repository>>, casbin_enforcer: Arc<CasbinEnforcer>) -> Self {
+        Self {
+            user_repository,
+            casbin_enforcer,
+        }
     }
 
+    ///Allows or denies an user to perform an action based on the user's privileges
+    ///
     /// # Errors
     ///
     /// Will return an error if:
-    ///
-    /// - There is not any user with the provided `UserId` (when the user id is some).
+    /// - There is no user id found in the request
+    /// - The user id is not found in the database
     /// - The user is not authorized to perform the action.
-    pub async fn authorize(&self, action: ACTION, maybe_user_id: Option<UserId>) -> Result<(), ServiceError> {
-        match action {
-            ACTION::AddCategory
-            | ACTION::DeleteCategory
-            | ACTION::GetSettings
-            | ACTION::GetSettingsSecret
-            | ACTION::AddTag
-            | ACTION::DeleteTag => match maybe_user_id {
-                Some(user_id) => {
-                    let user = self.get_user(user_id).await?;
-
-                    if !user.administrator {
-                        return Err(ServiceError::Unauthorized);
-                    }
 
+    pub async fn authorize(&self, action: ACTION, maybe_user_id: Option<UserId>) -> std::result::Result<(), ServiceError> {
+        match maybe_user_id {
+            Some(user_id) => {
+                // Checks if the user found in the request exists in the database
+                let user_guard = self.get_user(user_id).await?;
```

Comment:
Hi @mario-nt, why is this called `user_guard`?

Reply:
I thought it was a good name because it checks if the user exists in the database, if not, returns an error. I can change it if you think it is not appropriate.